feat: OAS validation, better error handling, action debugging

kanadgupta opened the pull request:
## What's Changed?

#### ... A lot! So here are the highlights:
- [x] feat: validate OAS before bundling (5620362)
  - This validates the OAS before we send it via API. Not sure if this will be controversial, maybe we can add this behind a flag so it isn't breaking? Open for discussion on this.
- [x] feat: add debugging (fe5555d)
  - Utilizes GitHub's [Action Debugger](https://github.com/actions/toolkit/blob/master/docs/action-debugging.md) so customers have a little more info to give us when they see issues
- [x] fix: properly parse our error messages (0a230bd)
  - Actually handles our new error message structure, closes #23
- [x] fix: error out if no file is found (c1358ac)
  - Adds an error case for if no file is found

#### As well as some housekeeping:
- [x] ci: add prettier to package.json (4297748)
- [x] ci: add back warnings for console.logs() (1705bf7)
- [x] docs: update some log messages (17c145a)
- [x] refactor: remove an unnecessary if block (7ba10ad)

kanadgupta committed:
ci: add prettier to package.json

This allows us to better utilize the prettier config locally!

kanadgupta committed:
fix: error out if no file is found

kanadgupta committed:
ci: add back warnings for console.logs()

We should use the core.info() statements instead!

Info: https://github.com/actions/toolkit/tree/master/packages/core#logging

kanadgupta committed:
feat: validate OAS before bundling

kanadgupta committed:
docs: update some log messages

Convert a console.log() to core.info() and update some language

kanadgupta committed:
refactor: remove an unnecessary if block

Now that we're validating the OAS beforehand, this is unnecessary.

Also removes a now-completed TODO :tada:

kanadgupta committed:
fix: properly parse our error messages

Refactors the parsing logic to handle the new error message structure, and also adds a few additional helpful error messages for certain errors.

Also slightly refactored the version logic so we can use version in one of the error messages.

kanadgupta committed:
feat: add debugging

GitHub Actions support debugging which may be useful for customers that have issues, so I added a bunch of debug statements. And no, they won't get logged by default!

Info: https://github.com/actions/toolkit/blob/master/docs/action-debugging.md

Also I added logic to the demo executor to better log these debug lines, and refactored some of the output logic.

erunion commented on main.js:
```diff
@@ -39,21 +48,27 @@ async function run() {
-    .then(generatedSwaggerString => {
+    .then(async generatedSwaggerString => {
       const oas = new OAS(generatedSwaggerString);
+      const validate = promisify(oas.validate);
```
Real promise support on this library is coming soon!

erunion commented on main.js:
```diff
@@ -105,7 +139,7 @@ async function run() {
     })
     .catch(err => {
-      core.setFailed(`There was an error finding or loading your OAS file.\n\n${err.message || err}`);
+      core.setFailed(`There was an error finding or loading your OpenAPI/Swagger file.\n\n${err.message || err}`);
```
rdme has this similar error message and I've always wanted it to be contextual to what happened so if we borked an OAS it'd never mention Swagger.

erunion reviewed (commented):
this all looks awesome.

do you also want to see if you can get the action working without having `node_modules/` stored in git? 🥺

kanadgupta committed:
deps: try commiting node_modules

Because of an error I'm seeing here: https://github.com/kanadgupta/metrotransit-nextrip-oas/runs/1261522969

kanadgupta commented:
> do you also want to see if you can get the action working without having `node_modules/` stored in git? 🥺

@erunion yeah, I don't love that either... [the GitHub docs](https://docs.github.com/en/free-pro-team@latest/actions/creating-actions/creating-a-javascript-action#commit-tag-and-push-your-action-to-github:~:text=As%20an%20alternative%20to%20checking%20in,into%20one%20file%20used%20for%20distribution.) suggest using [`@vercel/ncc`](https://github.com/vercel/ncc) instead of `node_modules/`, which would have us commit a dist file instead. Should we do that?

erunion commented:
@kanadgupta I'd be good with that or a Docker solution like in https://github.com/brpaz/action-semantic-release

https://github.com/brpaz/action-semantic-release/blob/master/Dockerfile
https://github.com/brpaz/action-semantic-release/blob/master/action.yml#L13-L15
https://github.com/brpaz/action-semantic-release/blob/master/entrypoint.sh

Any solution is probably out of scope for this PR, but whatever you want to do is cool with me.

kanadgupta committed:
fix: move split into try catch

kanadgupta committed:
docs: improve some error messages, add a link to our docs

kanadgupta committed:
fix: sanitize all API keys, etc. before writing to stdout

kanadgupta committed:
chore: remove ReadMe API key entirely from error message

sanitizing it won't really help in this case! having a placeholder should make it clearer

kanadgupta committed:
chore: oops

we can't use the function before it's defined!

kanadgupta committed:
Merge branch 'master' into fix/error-handling

mjcuva reviewed (approved):
I haven't tested, but this all seems good to me!
